cpp/src/arrow/dataset/file_json.cc

```diff
@@ -0,0 +1,414 @@
+  if (!pool) pool = default_memory_pool();
+
+  auto full_buffer = std::make_shared<Buffer>(data);
+  std::shared_ptr<Buffer> buffer, partial;
+  auto chunker = json::MakeChunker(parse_options);
+  RETURN_NOT_OK(chunker->Process(full_buffer, &buffer, &partial));
+
+  std::unique_ptr<json::BlockParser> parser;
+  RETURN_NOT_OK(json::BlockParser::Make(pool, parse_options, &parser));
+  RETURN_NOT_OK(parser->Parse(buffer));
+  std::shared_ptr<Array> parsed;
+  RETURN_NOT_OK(parser->Finish(&parsed));
+
+  return checked_pointer_cast<StructType>(parsed->type());
+}
+
+Result<std::shared_ptr<Schema>> ParseToSchema(std::string_view data,
+                                              const json::ParseOptions& parse_options,
+                                              MemoryPool* pool) {
+  ARROW_ASSIGN_OR_RAISE(auto type, ParseToStructType(data, parse_options, pool));
+  return ::arrow::schema(type->fields());
+}
+
+// Converts a FieldPath to a FieldRef consisting exclusively of field names.
+//
+// The resulting FieldRef can be used to lookup the corresponding field in any schema
+// regardless of missing/unordered fields. The input path is assumed to be valid for the
+// given schema.
+FieldRef ToUniversalRef(const FieldPath& path, const Schema& schema) {
+  std::vector<FieldRef> refs;
+  refs.reserve(path.indices().size());
+
+  const FieldVector* fields = &schema.fields();
+  for (auto it = path.begin(); it != path.end(); ++it) {
+    DCHECK_LT(*it, static_cast<int>(fields->size()));
+    const auto& child_field = *(*fields)[*it];
+    refs.push_back(FieldRef(child_field.name()));
+    if (it + 1 != path.end()) {
+      auto&& child_type = checked_cast<const StructType&>(*child_field.type());
+      fields = &child_type.fields();
+    }
+  }
+
+  return refs.empty()       ? FieldRef()
+         : refs.size() == 1 ? refs[0]
+                            : FieldRef(std::move(refs));
+}
+
+int TopLevelIndex(const FieldRef& ref, const Schema& schema) {
+  if (const auto* name = ref.name()) {
+    return schema.GetFieldIndex(*name);
+  } else if (const auto* path = ref.field_path()) {
+    DCHECK(!path->empty());
+    return (*path)[0];
+  }
+  const auto* nested_refs = ref.nested_refs();
+  DCHECK(nested_refs && !nested_refs->empty());
+  return TopLevelIndex((*nested_refs)[0], schema);
+}
+
+// Make a new schema consisting only of the top-level fields in the dataset schema that:
+//  (a) Have children that require materialization
+//  (b) Have children present in `physical_schema`
+//
+// The resulting schema can be used in reader instantiation to ignore unused fields. Note
+// that `physical_schema` is only of structural importance and its data types are ignored
+// when constructing the final schema.
+Result<std::shared_ptr<Schema>> GetPartialSchema(const ScanOptions& scan_options,
+                                                 const Schema& physical_schema) {
```

Comment:
This "shouldn't" be needed in the new scanner.  I don't remember if it's needed in the old path or not.  If it is then we need to keep it unfortunately.  We can re-examine later when adding support for the new path.

Reply:
Yeah, from what I can tell, the new scanner does a lot of the legwork here (as far as top-level fields, at least). For the old path, this was what was required to get the common format tests to pass.